public/app/plugins/datasource/cloudwatch/migrations/metricQueryMigrations.ts

```diff
@@ -0,0 +1,35 @@
+import { config } from '@grafana/runtime';
+
+import { CloudWatchMetricsQuery } from '../types';
+
+// Call this function to migrate queries from within the plugin.
+export function migrateMetricQuery(query: CloudWatchMetricsQuery): CloudWatchMetricsQuery {
+  //add metric query migrations here
+  const migratedQuery = migrateAliasPatterns(query);
+  return migratedQuery;
+}
+
+const aliasPatterns: Record<string, string> = {
+  metric: `PROP('MetricName')`,
+  namespace: `PROP('Namespace')`,
+  period: `PROP('Period')`,
+  region: `PROP('Region')`,
+  stat: `PROP('Stat')`,
+  label: `LABEL`,
+};
+
+export function migrateAliasPatterns(query: CloudWatchMetricsQuery): CloudWatchMetricsQuery {
```

Comment:
do we need this migration if this is done already in the backend? https://github.com/grafana/grafana/pull/48555. I mean, this will migrate the query so it will no longer require the migration in the backend but isn't it a pain to maintain the same code twice?

Reply:
Ah, I guess this is to display the new format in the UI from a previous query. It seems that this is also necessary :thinking:

Food for thought but maybe it make sense for the backend to have a migration endpoint (`POST /migrate`) so the code is not duplicated.

Reply:
Yeah. And another problem with a backend migration is that it can only do the migration on the fly - it cannot change the actual query. This however is possible in the frontend (see https://github.com/grafana/grafana/blob/cloudwatch/dynamic-labels-frontend-migration/public/app/plugins/datasource/cloudwatch/components/usePreparedMetricsQuery.ts#L43).

Having one, standardized way of doing migrations across frontend and backend would be really nice. https://github.com/grafana/grafana/issues/31318